File: SeparatingAxisTest/Vector.m
//
//  Vector.m
//  SeparatingAxisTest
//
//  Created by Najati Imam on 7/26/12.
//  Copyright (c) 2012 Cyrus Innovation. All rights reserved.
//

#import "Vector.h"

@implementation Vector

+(Vector*) x:(float) x y:(float) y {
    return [[Vector alloc] initWithX:x andY:y]; 
}

- (id) init {
    if (self = [super init]) {
        coords = malloc(2 * sizeof(float));
        coords[0] = 0;
        coords[1] = 0;
    }
    return self;
}

- (id) initWithX: (float) x andY: (float) y {
    if (self = [super init]) {
        coords = malloc(2 * sizeof(float));
        coords[0] = x;
        coords[1] = y;
    }
    return self;
}

- (void) dealloc {
    free(coords);
}

-(float) x { return coords[0]; }
-(float) y { return coords[1]; }

-(float) dot:(Vector *)vector {
    return coords[0]*(vector->coords[0]) + coords[1]*(vector->coords[1]);
}

-(float) lengthSquared {
    return coords[0]*coords[0] + coords[1]*coords[1];
}

-(Vector*) minus:(Vector*) vector {
    return [Vector x:coords[0] - (vector->coords[0]) y: coords[1] - (vector->coords[1])];
}

-(void) normalize {
    float length = sqrtf(coords[0]*coords[0] + coords[1]*coords[1]);
    coords[0] = coords[0]/length;
    coords[1] = coords[1]/length;
}

-(void) flop {
    float x = coords[0];
    coords[0] = -coords[1];
    coords[1] = x;
}

-(void) copyFrom:(Vector*) vector {
    coords[0] = (vector->coords[0]);
    coords[1] = (vector->coords[1]);
}

-(void) scaleBy:(float) scale {
    coords[0] = coords[0] * scale;
    coords[1] = coords[1] * scale;
}

@end

File: SeparatingAxisTest/Vector.c
//
//  Vector.m
//  SeparatingAxisTest
//
//  Created by Najati Imam on 7/26/12.
//  Copyright (c) 2012 Cyrus Innovation. All rights reserved.
//

#import "Vector.h"

#import "math.h"

Vector vector_from(float x, float y) {
    Vector v;
    v.x = x;
    v.y = y;
    return v; 
}

Vector vector_copy(Vector b) {
    Vector v;
    v.x = b.x;
    v.y = b.y;
    return v;
}

Vector vector_subtract(Vector minuend, Vector subtrahend) {
    Vector difference;
    difference.x = minuend.x - subtrahend.x;
    difference.y = minuend.y - subtrahend.y;
    return difference;
}

Vector vector_normal(Vector v) {
    vector_normalize(&v);
    return v;
}


float vector_dot(Vector a, Vector b) {
    return a.x*b.x + a.y*b.y;
}

float vector_length_squared(Vector v) {
    return v.x*v.x + v.y*v.y;
}


void vector_scale(Vector *v, float scale) {
    (*v).x = (*v).x * scale;
    (*v).y = (*v).y * scale;
}

void vector_normalize(Vector *v) {
    float length = sqrtf((*v).x*(*v).x + (*v).y*(*v).y);
    (*v).x = (*v).x/length;
    (*v).y = (*v).y/length;
}

void vector_flop(Vector *v) {
    float x = (*v).x;
    (*v).x = -(*v).y;
    (*v).y = x;
}
